### PR TITLE
Use ShelleyNTC for version check in node queries

### PR DESCRIPTION
**Pull request by carbolymer**
# Changelog

```yaml
- description: |
    Use ShelleyNTC for version check in node queries
# uncomment types applicable to the change:
  type:
   - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
   - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Fixes https://github.com/input-output-hk/cardano-api/issues/391

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/IPC/Monad.hs [hunk 1]
```diff
@@ -86,21 +121,121 @@ setupLocalStateQueryExpr waitDone mPointVar' resultVar' ntcVersion f =
...
+  case isQuerySupprted shelleyNtcVersion q of
+    Left e -> pure $ Left e
+    Right isShelleyNtcSupported
+      | ntcVersion >= minNtcVersion && isShelleyNtcSupported ->
+          fmap Right . LocalStateQueryExpr . ReaderT $ \_ -> ContT $ \f -> pure $
+            Net.Query.SendMsgQuery q $
+              Net.Query.ClientStQuerying
+              { Net.Query.recvMsgResult = f
+              }
+      | otherwise -> pure . Left $ UnsupportedNtcVersionError minNtcVersion ntcVersion
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/IPC/Monad.hs:172]: * [ ] TODO: include shelley NTC version info and/or use separate error for it

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
I haven't look at this yet but I'm prematurely requesting changes as this needs to be carefully considered.

**Review by amesgen — COMMENTED** [review]
The new logic here looks correct to me.

We also talked about this in the Consensus Office Hours today, and came to the conclusion that we can avoid `cardano-api` having to implement this more complicated logic by a simple change in Consensus, see #863 and #862.

**Comment by carbolymer** [comment]
Thanks @amesgen ! I think I prefer to rely on the consensus change and your assumption:
```
In order to reduce the number of possible version combinations, we currently follow the convention that all `ShelleyNodeToClientVersion`s in one `CardanoNodeToClientVersionX` are equal.
```
this way we can just use `NodeToClientVersion` and avoid introducing redundant code.
@Jimbo4350 thoughts?

**Comment by carbolymer** [comment]
Closing because of: https://github.com/IntersectMBO/cardano-api/pull/403#issuecomment-1880960379

**Pull request closed without merging**